### PR TITLE
Add RAP (BIP47)

### PR DESCRIPTION
**Pull request by HysMagus**
ZCOIN-64 Store the user's 10 derived incoming addresses in a new wallet.bip47 file
ZCOIN-62 Update to newly released Zcoin core version 13.8.5
ZCOIN-57 Upgrade the core codebase to 13.8.4
ZCOIN-56 Payment Codes Screen: Make the history list dynamic
ZCOIN-52 Transaction History: Show the Type as 'Payment Code' if a code was used to send or receive
ZCOIN-48 Add newly used Payment Codes to the new Payment Codes Address Book
ZCOIN-43 Test whether a newly created notification transaction is valid
ZCOIN-41 Create a Notification Transaction using receiver's masked payment code
ZCOIN-39 Masking: Calculate and handle the values needed for the second arg of the masking function
ZCOIN-38 Send Screen enhancements and modifications
ZCOIN-37 Data Model for storage of Notification Transactions, and the Qt display table
ZCOIN-36 Mask the Sender's Payment Code
ZCOIN-35 Qt > Payment Codes tab: QR code for scanning a user's own payment code
ZCOIN-34 Qt > Payment Codes tab: Implement the new tab, and display the user's code & copy button
ZCOIN-33 Create X and O inputs for masking a payment code
ZCOIN-29 Research current BIP-32 (HD) implementation in Zcoin source code
ZCOIN-27 Create Payment Code for Receiver
ZCOIN-22 Qt > Send: When pasting a string into the destination field, automatically determine if is a standard Zcoin address or a Payment Code
ZCOIN-20 RPC: sendtoPcode command
ZCOIN-19 RPC: createPcodenotification command
ZCOIN-18 RPC: validatePcode command
ZCOIN-17 RPC: importPcode command
ZCOIN-16 RPC: listPcodes command
ZCOIN-15 RPC: listreceivedbyPcode command
ZCOIN-14 RPC: getreceivedbyPcode command
ZCOIN-13 RPC: getbulkPcode command
ZCOIN-12 RPC: getMyPaymentCode command

This Pull Request covers adding Payment Codes Sending, Receiving, Sigma Spends to Payment Codes, Notification Transaction Prompts, wallet.bip47 wallet file (holds handshakes for BIP47), and UI Elements for Zcoin BIP47 UI.

Note: This PR is made on behalf of @arcadiamediagroup and not me as an individual, work reflected here is not my own but that of my team @decentralizedsolutions , @top1st , @gongfeigithub , @pvc2104 (@phamvancam2104) , and @melissamouton . 

Edit: Add Cam's other Github

### CODE REVIEW

File: src/bip47-test.cpp [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+//            b58key.ToString().c_str(),
+//            derive.prv.c_str());
+//
+//    CBitcoinExtKey b58keyDecodeCheck(derive.prv);
+//    CExtKey checkKey = b58keyDecodeCheck.GetKey();
+//    assert(checkKey == key);
+//
+//    printf("Test Passed\n");
+
+//    std::string strHexMaster = "b7b8706d714d9166e66e7ed5b3c61048";
```
Comment (akshaynexus) [inline src/bip47-test.cpp]: Any reason why the comment code is there in this file?Better to remove unused code if its not required

File: src/bip47-test.cpp [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+
+
+//    assert(key.VerifyPubKey(pubKey));
+
+
+//    CBitcoinAddress addr(strAddress);
+//    CKeyID keyId;
+//    addr.GetKeyID(keyId);
+
+//    pwalletMain->GetKey(hdChain.masterKeyID, key);
```
Comment (akshaynexus) [inline src/bip47-test.cpp]: Same question about this commented out code

File: src/bip47-test.cpp (hunk 1 above)
Comment (HysMagus) [reply src/bip47-test.cpp]: Hi Akshay! Sorry for the delayed response, it's not super necessary and can be removed

File: src/bip47-test.cpp (hunk 2 above)
Comment (HysMagus) [reply src/bip47-test.cpp]: Removed

File: src/bip47-test.cpp (hunk 1 above)
Comment (reubenyap) [reply src/bip47-test.cpp]: I believe this was cleaned up significantly in the following commits.

File: src/bip47/Bip47ChannelAddress.h [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    std::vector<unsigned char> pubKeyHash ;
+    
+public:
+    Bip47ChannelAddress() ;
+    Bip47ChannelAddress(CExtPubKey &cKey, int child) ;
+    std::vector<unsigned char>& getPubKey() ;
+
+    std::vector<unsigned char>& getPubKeyHash() ;
+
+    String getAddressString() ;
```
Comment (levonpetrosyan93) [inline src/bip47/Bip47ChannelAddress.h]: this function is not used anywhere, if we dont need this, better to remove it.

File: src/bip47/Bip47ChannelAddress.h [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+    String getPrivateKeyString() ;
```
Comment (levonpetrosyan93) [inline src/bip47/Bip47ChannelAddress.h]: this one also not used

File: src/Makefile.am [hunk 5]
```diff
@@ -126,7 +126,7 @@ TESTS =
 BENCHMARKS =
 
 if BUILD_BITCOIND
-  bin_PROGRAMS += zcoind
+  bin_PROGRAMS += zcoind testbip47
```
Comment (levonpetrosyan93) [inline src/Makefile.am]: I think better not to create new binary for bip47 testing, good to include it at test_bitcoin binary and use boost testing for it.

File: src/wallet/walletdb.h [hunk 6]
```diff
@@ -107,6 +109,51 @@ class CHDChain
     }
 };
 
+/**
+ * CBip47HDChain
+ */
+
+/* simple HD chain data model */
+class CBip47HDChain
```
Comment (levonpetrosyan93) [inline src/wallet/walletdb.h]: why do we need this?
looks this is not used anywhere.

File: src/wallet/walletdb.h [hunk 7]
```diff
@@ -298,6 +345,24 @@ class CWalletDB : public CDB
     bool WriteHDChain(const CHDChain& chain);
     bool WriteMnemonic(const MnemonicContainer& mnContainer);
 
+    //! write the bip47hdchain model (external chain child index counter)
+    bool WriteCBip47HDChain(const CBip47HDChain& bip47chain);
```
Comment (levonpetrosyan93) [inline src/wallet/walletdb.h]: don't need this also

File: src/wallet/walletdb.h [hunk 8]
```diff
@@ -298,6 +345,24 @@ class CWalletDB : public CDB
...
+    
+    // @bip47 channel data write
+    bool WriteBip47PaymentChannel(const Bip47PaymentChannel& pchannel, const string& channelId);
+    
+    void ListBip47PaymentChannel(std::map <string, Bip47PaymentChannel> &mPchannels);
+    
+    bool WritePaymentChannelData();
```
Comment (levonpetrosyan93) [inline src/wallet/walletdb.h]: this function is not defined anywhere, and also not used.

File: src/wallet/wallet.h [hunk 9]
```diff
@@ -88,6 +91,9 @@ const uint32_t BIP44_MINT_INDEX = 0x2;
 #ifdef ENABLE_EXODUS
 const uint32_t BIP44_EXODUS_MINT_INDEX = 0x3;
 #endif
+const uint32_t BIP47_INDEX = 0x2F;
+const uint32_t BIP47_TEST_INDEX = 0x1;   // https://github.com/satoshilabs/slips/blob/master/slip-0044.md#registered-coin-types lgtm [cpp/unused-static-variable] 
+const uint32_t BIP47_ZCOIN_INDEX = 0x88; // https://github.com/satoshilabs/slips/blob/master/slip-0044.md#registered-coin-types lgtm [cpp/unused-static-variable]
```
Comment (levonpetrosyan93) [inline src/wallet/wallet.h]: why not to remove above this two unused declarations ?

File: src/wallet/wallet.h [hunk 10]
```diff
@@ -690,9 +696,11 @@ class CWallet : public CCryptoKeyStore, public CValidationInterface
 
     bool fFileBacked;
     std::string strWalletFile;
+    static std::string bip47WalletFile;
 
     std::set<int64_t> setKeyPool;
     std::map<CKeyID, CKeyMetadata> mapKeyMetadata;
+    std::map<CKeyID, CKeyMetadata> Bip47mapKeyMetadata;
```
Comment (levonpetrosyan93) [inline src/wallet/wallet.h]: this one is also not being used.

File: src/bip47/Bip47Util.cpp [hunk 11]
```diff
@@ -0,0 +1,177 @@
...
+            op_data = data;
+            return true;
+        } 
+        
+    }
+}
+
+bool BIP47Util::getPaymentCodeInNotificationTransaction(vector<unsigned char> privKeyBytes, CTransaction tx, PaymentCode &paymentCode) {
+    // tx.vin[0].scriptSig
+    CWalletTx wtx(pwalletMain, tx);
```
Comment (levonpetrosyan93) [inline src/bip47/Bip47Util.cpp]: I think no need to create additional object here, you can get all data you need below from `CTransaction tx`

**Comment by HysMagus** [comment]
Now that these changes have gone in, I'll be doing some functional testing on testnet tomorrow to ensure nothing broke in the process :)

**Comment by levonpetrosyan93** [comment]
Looks we have some Ui issues, on "Receiving addresses" window both regular addresses and Bip47Receive are at "Zcoin Addresses" tab, "Payment Codes" tab is empty. I push "New" button, when "Payment Codes" tab is active, one more address is being added at "Zcoin Addresses" tab. Also "Close" button and "X" are not closing the window, when  "Payment Codes" tab is active. Issue with resizing also exists, when I resize the window, widget inside it is not being resized.
Looks same is at "Sending addresses"window. 
b.t.w I am using Ubuntu 18.04.4 LTS.

File: src/wallet/wallet.h [hunk 12]
```diff
@@ -1215,6 +1227,72 @@ class CWallet : public CCryptoKeyStore, public CValidationInterface
...
+    
+    std::string getNotifiactionAddress();
+
+    std::string getPaymentCode();
+    std::string getPaymentCodeForAddress(std::string address);
+    
+    void deriveBip47Accounts(std::vector<unsigned char> hd_seed);
+    void deriveBip47Accounts(CExtKey masterKey);
+    
+    bool importBip47PaymentChannelData();
```
Comment (levonpetrosyan93) [inline src/wallet/wallet.h]: looks this functions is not being used

File: src/wallet/wallet.h [hunk 13]
```diff
@@ -1215,6 +1225,71 @@ class CWallet : public CCryptoKeyStore, public CValidationInterface
...
+    bool isNotificationTransaction(CTransaction tx);
+    bool isNotificationTransactionSent(string pcodestr);
+    bool isToBIP47Address(CTransaction tx);
+    bool generateBip47SeedMaster(vector<unsigned char> &seedmaster);
+    bool saveBip47SeedMaster(vector<unsigned char> seedmaster);
+    bool loadBip47SeedMaster(vector<unsigned char>& seedmaster);
+    PaymentCode getPaymentCodeInNotificationTransaction(CTransaction tx);
+    
+    CBitcoinAddress getAddressOfReceived(CTransaction tx);
+    CBitcoinAddress getAddressOfSent(CTransaction tx);
```
Comment (levonpetrosyan93) [inline src/wallet/wallet.h]: this  function is not used

**Comment by reubenyap** [comment]
Closing this as this has been ported to core 14 via #834

**Pull request closed without merging**